lib/create_response.js

```diff
@@ -16,17 +18,58 @@ const { STATUS_CODES } = require('http')
 const responseStatusCodesWithoutBody = [204, 205, 304]
 
 /**
- * @param {IncomingMessage} message
+ * @param {import('node:http').IncomingMessage} message
  */
 function createResponse(message) {
+  // https://github.com/Uzlopak/undici/blob/main/lib/fetch/index.js#L2031
+  const decoders = []
+  const codings =
+    message.headers['content-encoding']
+      ?.toLowerCase()
+      .split(',')
+      .map(x => x.trim())
+      .reverse() || []
+  for (const coding of codings) {
+    if (coding === 'gzip' || coding === 'x-gzip') {
+      decoders.push(
+        zlib.createGunzip({
+          flush: zlib.constants.Z_SYNC_FLUSH,
+          finishFlush: zlib.constants.Z_SYNC_FLUSH,
+        }),
+      )
+    } else if (coding === 'deflate') {
+      decoders.push(zlib.createInflate())
+    } else if (coding === 'br') {
+      decoders.push(zlib.createBrotliDecompress())
+    } else {
+      decoders.length = 0
+      break
+    }
+  }
+
+  const chunks = []
   const responseBodyOrNull = responseStatusCodesWithoutBody.includes(
     message.statusCode,
   )
     ? null
     : new ReadableStream({
         start(controller) {
-          message.on('data', chunk => controller.enqueue(chunk))
-          message.on('end', () => controller.close())
+          message.on('data', chunk => chunks.push(chunk))
+          message.on('end', () => {
+            pipeline(
+              Readable.from(chunks),
+              ...decoders,
+              async function* (source) {
+                for await (const chunk of source) {
+                  yield controller.enqueue(chunk)
+                }
+              },
+              e => {
+                if (e) throw e
+                controller.close()
+              },
```

Comment:
Will this not result in the error being swallowed?

why not passing the error to the controller?

```suggestion
             controller.error,
```

Or do you need to call then explicitly controller.close after calling error?

```suggestion
              e => {
                controller.error(e)
                controller.close()
              },
```

But I am not sure if this is right.

Reply:
I honestly am new to webstreams

Reply:
Sorry for the confusion, this is a callback for the pipeline, I rewrite it for better readablity and use the controller.error instead of throwing it.